src/pybind/mgr/object_format.py

```diff
@@ -70,3 +91,131 @@ def to_yaml(obj: Any, *, many: bool = False, cls: Any = None) -> Any:
+    Callable[..., Dict[Any, Any]],
+    Callable[..., List[Any]],
+]
+
+
+class ErrorResponseBase(Exception):
+    """An exception that can directly be converted to a mgr reponse."""
+
+    def format_response(self) -> Tuple[int, str, str]:
+        raise NotImplementedError()
+
+
+class UnknownFormat(ErrorResponseBase):
+    """Raised if the format name is unexpected.
+    This can help distinguish typos from formats that are known but
+    not implemented.
+    """
+
+    def __init__(self, format_name: str) -> None:
+        self.format_name = format_name
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return -errno.EINVAL, "", f"Unknown format name: {self.format_name}"
+
+
+class UnsupportedFormat(ErrorResponseBase):
+    """Raised if the format name does not correspond to any valid
+    conversion functions.
+    """
+
+    def __init__(self, format_name: str) -> None:
+        self.format_name = format_name
+
+    def format_response(self) -> Tuple[int, str, str]:
+        return -errno.EINVAL, "", f"Unsupported format: {self.format_name}"
+
+
+class Formatter(Protocol):
+    def valid_formats(self) -> Iterable[str]:
+        ...
+
+
+class StandardFormatter:
+    """A standard formatter that can take basic objects types and convert them
+    to JSON and/or YAML.
+    """
+
+    def __init__(self, indent: Optional[int] = 2, many: bool = False) -> None:
+        self.indent = indent
+        self.many = many
+        self._valid_formats = set(
+            v.value for k, v in Format.__members__.items()
+        )
+
+    def valid_formats(self) -> Iterable[str]:
+        return self._valid_formats
+
+    def format_json(self, obj: Any) -> str:
+        return to_json(obj, indent=self.indent, many=self.many)
+
+    format_json_pretty = format_json
+
+    def format_yaml(self, obj: Any) -> str:
+        return to_yaml(obj, many=self.many)
+
+
+def formatted(
+    obj: Any,
+    format_req: Optional[str] = None,
+    formatter: Optional[Formatter] = None,
+) -> str:
+    """Return the content of the object `obj` formatted into the
+    given serialization format given by `format_req` and applied by
+    the optional formatter.
+    """
+    if format_req is None:
+        format_req = "json"  # default to json
+    if formatter is None:
+        formatter = StandardFormatter()
+    if format_req not in formatter.valid_formats():
+        raise UnknownFormat(format_req)
+    req = str(format_req).replace("-", "_")
+    ffunc = getattr(formatter, f"format_{req}", None)
+    if ffunc is None:
+        raise UnsupportedFormat(format_req)
+    return ffunc(obj)
+
+
+def return_value(obj: Any) -> int:
+    """Fetch a return value for the mgr based on an optional
+    `return_value` method on the object.
+    """
+    rvfunc = getattr(obj, "return_value", None)
+    if rvfunc is None:
+        return 0
+    return rvfunc()
+
+
+def _get_requested_format(f: ObjectResponseFuncType, kw: Dict[str, Any]) -> str:
```

Comment:
As the `todo` below indicates the idea was to only remove `format`from the keyword iff the function `f` does not make use of the `format` argument. We need f to get f's signature to make that decision.
However, its clear that I didn't implement this for these patches. It turned out unnecessary in the functions I was working with so I left the work for another time. I suppose since this is a private function we can claim it's a YAGNI condition. I'd prefer to leave it as-is but I don't feel strongly about it and would be open to simplifying it.

Reply:
We can leave it, was just curious why it was being passed since it wasn't used directly.